imod/prepare/topsystem/allocation.py

```diff
@@ -0,0 +1,329 @@
+"""
+This module contains all kinds of utilities to prepare rivers
+"""
+
+from enum import Enum
+from typing import Optional
+
+from imod.prepare.layer import (
+    create_layered_top,
+    get_upper_active_grid_cells,
+    get_upper_active_layer_number,
+)
+from imod.schemata import DimsSchema
+from imod.typing import GridDataArray
+from imod.util.dims import enforced_dim_order
+
+
+class ALLOCATION_OPTION(Enum):
+    """
+    Enumerator for allocation settings. Numbers match the IDEFLAYER options in
+    iMOD 5.6.
+    """
+
+    stage_to_riv_bot = 0
+    first_active_to_riv_bot = -1
+    first_active_to_riv_bot__drn = 1
+    at_elevation = 2
+    at_first_active = 9  # Not an iMOD 5.6 option
+
+
+PLANAR_GRID = (
+    DimsSchema("time", "y", "x")
+    | DimsSchema("y", "x")
+    | DimsSchema("time", "{face_dim}")
+    | DimsSchema("{face_dim}")
+)
+
+
+def allocate_riv_cells(
+    allocation_option: ALLOCATION_OPTION,
+    active: GridDataArray,
+    top: GridDataArray,
+    bottom: GridDataArray,
+    stage: GridDataArray,
+    bottom_elevation: GridDataArray,
+) -> tuple[GridDataArray, Optional[GridDataArray]]:
+    match allocation_option:
+        case ALLOCATION_OPTION.stage_to_riv_bot:
+            return _allocate_cells__stage_to_riv_bot(
+                top, bottom, stage, bottom_elevation
+            )
+        case ALLOCATION_OPTION.first_active_to_riv_bot:
+            return _allocate_cells__first_active_to_riv_bot(
+                active, top, bottom, bottom_elevation
+            )
+        case ALLOCATION_OPTION.first_active_to_riv_bot__drn:
+            return _allocate_cells__first_active_to_riv_bot__drn(
+                active, top, bottom, stage, bottom_elevation
+            )
+        case ALLOCATION_OPTION.at_elevation:
+            return _allocate_cells__at_elevation(top, bottom, bottom_elevation)
+        case ALLOCATION_OPTION.at_first_active:
+            return _allocate_cells__at_first_active(active)
+
+
+def allocate_drn_cells(
+    allocation_option: ALLOCATION_OPTION,
+    active: GridDataArray,
+    top: GridDataArray,
+    bottom: GridDataArray,
+    elevation: GridDataArray,
+) -> GridDataArray:
+    match allocation_option:
+        case ALLOCATION_OPTION.at_elevation:
+            return _allocate_cells__at_elevation(top, bottom, elevation)[0]
+        case ALLOCATION_OPTION.at_first_active:
+            return _allocate_cells__at_first_active(active)[0]
+        case _:
+            raise ValueError(
+                "Received incompatible setting for drains, only"
+                f"'{ALLOCATION_OPTION.at_elevation.name}' and"
+                f"'{ALLOCATION_OPTION.at_first_active.name}' supported."
+                f"got: '{allocation_option.name}'"
+            )
+
+
+def allocate_ghb_cells(
+    allocation_option: ALLOCATION_OPTION,
+    active: GridDataArray,
+    top: GridDataArray,
+    bottom: GridDataArray,
+    head: GridDataArray,
+) -> GridDataArray:
```

Comment:
for all these allocation functions, input arrays like "top" and "bottom" should all be xr.DataArray or xu.UgridDataArray but they should (I assume) be all of the same type. If so, could you do some validation? you can probably use the same validation function for all allocators.

Reply:
Not necessarily true:

You can supply ``top`` and ``bottom`` as scalar xr.DataArray and 1D xr.DataArray with layer dimension and provide unstructured grids to the other variables.